File: src/bin/io/RandomAccessFile.java
package bin.io;

import java.io.Closeable;
import java.io.EOFException;
import java.io.IOException;

/**
 * @author Bin
 */
public interface RandomAccessFile extends Closeable {

    void write(int value) throws IOException;

    void write(byte[] data) throws IOException;

    void write(byte[] data, int off, int len) throws IOException;

    int read() throws IOException;

    int read(byte[] data) throws IOException;

    int read(byte[] data, int off, int len) throws IOException;

    void readFully(byte[] data) throws IOException;

    void readFully(byte[] data, int off, int len) throws IOException;

    long length() throws IOException;

    void setLength(long newLength) throws IOException;

    void seek(long pos) throws IOException;

    int skipBytes(int n) throws IOException;

    long getFilePointer() throws IOException;

    String getName();

    RandomAccessFile getAnotherInSameParent(String name) throws IOException;

    RandomAccessFile newSameInstance() throws IOException;

    RandomAccessFile newFragment(long offset, long length) throws IOException;

    void flush() throws IOException;

    boolean isClosed();

    default void writeByte(byte b) throws IOException {
        write(b);
    }

    default void writeUShort(int i) throws IOException {
        write(i & 0xFF);
        write(i >>> 8 & 0xFF);
    }

    default void writeShort(short i) throws IOException {
        writeUShort(i);
    }

    default void writeChar(char c) throws IOException {
        writeUShort(c);
    }

    default void writeInt(int i) throws IOException {
        write(i & 0xFF);
        write(i >>> 8 & 0xFF);
        write(i >>> 16 & 0xFF);
        write(i >>> 24 & 0xFF);
    }

    default void writeLong(long l) throws IOException {
        write((int) (l & 0xFF));
        write((int) (l >>> 8 & 0xFF));
        write((int) (l >>> 16 & 0xFF));
        write((int) (l >>> 24 & 0xFF));
        write((int) (l >>> 32 & 0xFF));
        write((int) (l >>> 40 & 0xFF));
        write((int) (l >>> 48 & 0xFF));
        write((int) (l >>> 56 & 0xFF));
    }

    default byte readByte() throws IOException {
        int ret = read();
        if (ret == -1) {
            throw new EOFException();
        }
        return (byte) ret;
    }

    default int readUShort() throws IOException {
        return readByte() & 0xFF | (readByte() & 0xFF) << 8;
    }

    default short readShort() throws IOException {
        return (short) readUShort();
    }

    default char readChar() throws IOException {
        return (char) readUShort();
    }

    default int readInt() throws IOException {
        return readByte() & 0xFF | (readByte() & 0xFF) << 8 | (readByte() & 0xFF) << 16 | (readByte() & 0xFF) << 24;
    }

    default long readLong() throws IOException {
        return readByte() & 0xFFL | (readByte() & 0xFFL) << 8 | (readByte() & 0xFFL) << 16 | (readByte() & 0xFFL) << 24
                | (readByte() & 0xFFL) << 32 | (readByte() & 0xFFL) << 40 | (readByte() & 0xFFL) << 48 | (readByte() & 0xFFL) << 56;
    }


}

File: Android/app/src/main/java/bin/io/RandomAccessFile.kt
package bin.io

import java.io.Closeable
import java.io.EOFException
import java.io.IOException

/**
 * @author Bin
 */
interface RandomAccessFile : Closeable {
    @Throws(IOException::class)
    fun write(value: Int)

    @Throws(IOException::class)
    fun write(data: ByteArray)

    @Throws(IOException::class)
    fun write(data: ByteArray, off: Int, len: Int)

    @Throws(IOException::class)
    fun read(): Int

    @Throws(IOException::class)
    fun read(data: ByteArray): Int

    @Throws(IOException::class)
    fun read(data: ByteArray, off: Int, len: Int): Int

    @Throws(IOException::class)
    fun readFully(data: ByteArray)

    @Throws(IOException::class)
    fun readFully(data: ByteArray, off: Int, len: Int)

    @Throws(IOException::class)
    fun length(): Long

    @Throws(IOException::class)
    fun setLength(newLength: Long)

    @Throws(IOException::class)
    fun seek(pos: Long)

    @Throws(IOException::class)
    fun skipBytes(n: Int): Int

    @get:Throws(IOException::class)
    val filePointer: Long
    val name: String?

    @Throws(IOException::class)
    fun getAnotherInSameParent(name: String): RandomAccessFile

    @Throws(IOException::class)
    fun newSameInstance(): RandomAccessFile

    @Throws(IOException::class)
    fun newFragment(offset: Long, length: Long): RandomAccessFile

    @Throws(IOException::class)
    fun flush()
    val isClosed: Boolean

    @Throws(IOException::class)
    fun writeByte(b: Byte) {
        write(b.toInt())
    }

    @Throws(IOException::class)
    fun writeUShort(i: Int) {
        write(i and 0xFF)
        write(i ushr 8 and 0xFF)
    }

    @Throws(IOException::class)
    fun writeShort(i: Short) {
        writeUShort(i.toInt())
    }

    @Throws(IOException::class)
    fun writeChar(c: Char) {
        writeUShort(c.code)
    }

    @Throws(IOException::class)
    fun writeInt(i: Int) {
        write(i and 0xFF)
        write(i ushr 8 and 0xFF)
        write(i ushr 16 and 0xFF)
        write(i ushr 24 and 0xFF)
    }

    @Throws(IOException::class)
    fun writeLong(l: Long) {
        write((l and 0xFFL).toInt())
        write((l ushr 8 and 0xFFL).toInt())
        write((l ushr 16 and 0xFFL).toInt())
        write((l ushr 24 and 0xFFL).toInt())
        write((l ushr 32 and 0xFFL).toInt())
        write((l ushr 40 and 0xFFL).toInt())
        write((l ushr 48 and 0xFFL).toInt())
        write((l ushr 56 and 0xFFL).toInt())
    }

    @Throws(IOException::class)
    fun readByte(): Byte {
        val ret = read()
        if (ret == -1) {
            throw EOFException()
        }
        return ret.toByte()
    }

    @Throws(IOException::class)
    fun readUShort(): Int {
        return readByte().toInt() and 0xFF or (readByte().toInt() and 0xFF shl 8)
    }

    @Throws(IOException::class)
    fun readShort(): Short {
        return readUShort().toShort()
    }

    @Throws(IOException::class)
    fun readChar(): Char {
        return readUShort().toChar()
    }

    @Throws(IOException::class)
    fun readInt(): Int {
        return readByte().toInt() and 0xFF or (readByte().toInt() and 0xFF shl 8) or (readByte().toInt() and 0xFF shl 16) or (readByte().toInt() and 0xFF shl 24)
    }

    @Throws(IOException::class)
    fun readLong(): Long {
        return readByte().toLong() and 0xFFL or (readByte().toLong() and 0xFFL shl 8) or (readByte().toLong() and 0xFFL shl 16) or (readByte().toLong() and 0xFFL shl 24
                ) or (readByte().toLong() and 0xFFL shl 32) or (readByte().toLong() and 0xFFL shl 40) or (readByte().toLong() and 0xFFL shl 48) or (readByte().toLong() and 0xFFL shl 56)
    }
}